pkg/cni/server.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) 2021 Multus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package cni
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/invoke"
+	cnitypes "github.com/containernetworking/cni/pkg/types"
+	cnicurrent "github.com/containernetworking/cni/pkg/types/current"
+	"github.com/gorilla/mux"
+
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/config"
+	k8s "gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/k8sclient"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/logging"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/multus"
+	multustypes "gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/types"
+)
+
+// HandleCNIRequest is the CNI server handler function; it is invoked whenever
+// a CNI request is processed.
+func HandleCNIRequest(request *PodRequest) ([]byte, error) {
+	var result []byte
+	var err error
+
+	logging.Verbosef("%+v %s starting CNI request %+v", request, request.Command, request)
+	switch request.Command {
+	case "ADD":
+		result, err = request.cmdAdd()
+	case "DEL":
+		err = request.cmdDelete()
+	case "CHECK":
+		err = request.cmdCheck()
+	default:
+	}
+	logging.Verbosef("%s finished CNI request %+v, result: %q, err: %v", request.Command, request, string(result), err)
+	if err != nil {
+		// Prefix errors with request info for easier failure debugging
+		return nil, fmt.Errorf("%+v ERRORED: %v", request, err)
+	}
+	return result, nil
+}
```

Comment:
How do you plan I do this @s1061123 without adding a new struct ? How can I (on the multus-daemon side) know what to handle if I'm just using `skel.Args` ?

Reply:
... obviously having 3 endpoint handlers (instead of a single common one) would do it.

Reply:
If I design the code, I will do following step:
# I don't add error handling to simplify the process but of course, it is required.

1. (as you did,) CNI shim send CNI request through [type Request](https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/types.go#L51-L57)
1. based on https://github.com/containernetworking/cni/blob/master/pkg/skel/skel.go#L57-L154, wrote function that convert 'type Request' to 'skel.CmdArgs'. Let's name reqToCmdArgs() for instance. This function should return triplet, (cmd, CmdArgs, error).
1. At handleCNIRequest() in pkg/cni/server.go, after json.Unmarshal() to get Request structure, then call reqToCmdArgs() to get the cmd and CmdArgs.
1. Based on cmd, call one of CmdAdd/CmdDel/CmdCheck with CmdArgs.

This process does not introduce 'type Cmd' and 'type PodRequest'.

Reply:
I see. Now the question is: why is returning a triplet better than returning a struct that has those things embedded ? 

My issue w/ returning more than 2 arguments: I loose track of what they are. 

It's hard for me to use these things correctly, because what the attribute is is directly correlated to the order it is returned. I'd rather do that by name, which has meaning.

Reply:
I see both sides. So... returning a struct could be useful if it has a strong correlation with having to pass that object around between a bunch of different things. This does add some complexity though, and you have to look up the structure definition.

However, since we have named returned values in go code, we can always see those in the function definition, which does give them meaning. 

So I'd probably lean towards the triplet unless there's a case for tying the values together in a number of other places.

Reply:
> I see both sides. So... returning a struct could be useful if it has a strong correlation with having to pass that object around between a bunch of different things. This does add some complexity though, and you have to look up the structure definition.

... the same way I have to look up whatever the function's returning. 

> 
> However, since we have named returned values in go code, we can always see those in the function definition, which does give them meaning.

For that I have to navigate to the function's definition. I.e. the caller knows nothing about the returned types until he moves to the function's definition. 

Named return arguments are not a feature I would happily use. And I think it is not being used anywhere in the project.

> 
> So I'd probably lean towards the triplet unless there's a case for tying the values together in a number of other places.

I personally don't see how, but I will not choose this particular hill to die on. Whatever you guys say. I do think it reads worse, but I've already stated my opinion.

Reply:
> I personally don't see how, but I will not choose this particular hill to die on. Whatever you guys say. I do think it reads worse, but I've already stated my opinion.

I understand that you want not to use multiple return value for the golang code, however, the reason seems to be subjective, based on  your impression, not objective, hence we cannot agree on that because we don't have such impression.

Return multiple value is pretty usual language feature and not so unique feature that only in golang. As far as I know of, multiple return value (or similar language functionality) is supported not only in go, also in lisp, Standard ML, haskell, ocaml, ruby, python and rust. 

Many software projects written in golang uses multiple return value as usual as following:

- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/pods/helpers.go#L65
- https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/services/template.go#L297
- https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/net/pipe.go;l=113

If you suppose multiple values is the particular hill to die on, then these projects may be the one of them. If you want to change these culture, here is not the place to discuss about it. Please go appropriate places to discuss. 

Considering code volumes (less code is easy to maintain) and simplicity (i.e. less structure we need to maintain), we want to use multiple return value as usual golang does, without any structures.

Reply:
> If you suppose multiple values is the particular hill to die on, then these projects may be the one of them. If you want to change these culture, here is not the place to discuss about it. Please go appropriate places to discuss.

I don't think you understood me; I said "I will **not** choose this particular hill to die on. Whatever you guys say. ". This means that while I disagree with you, I'll comply. 

I'm waiting for the rest of the code to be reviewed before updating the PR.

Reply:
I suppose that the update (to incorporate with my comment) seems to be big and it also changes some codes not only in cni/server.go also types/conf.go and others. So could you please proceed your change first?

Reply:
Done - no more `PodRequest`, `command`, and now everything is passed via function parameters. 

Resolving this conversation.